src/goto-instrument/contracts/doc/user/contracts-memory-predicates.md

```diff
@@ -119,6 +120,117 @@ int foo()
 }
 ```
+## The __CPROVER_is_in_range predicate
+### Syntax
+
+```c
+bool __CPROVER_is_in_range(void *lb, void *p, void *ub);
+```
+
+This predicate holds if `lb`, `p` and `ub` are valid pointer within the same
+object and the pointers are ordered such that `lb <= p && p <= ub` holds.
+
+### Semantics
+In assertion contexts, the predicate checks the conditions described above.
+In assumption contexts, the predicate checks that `lb` and `ub` are valid pointers
+into the same object, and updates `p` using a side effect to be a non-deterministic
+pointer ranging between `lb` and `ub`.
+
+## User defined memory predicates
+
+Users can write their own memory predicates based on the two primitives above.
+`__CPROVER_is_fresh` allows to specify pointer validity and separation and
+`__CPROVER_pointer_in_range` allows to specify aliasing constraints.
+
+For instance, one could write a predicate defining linked lists of at most `len`
+elements as follows:
+
+```c
+typedef struct list_t
+{
+  int value;
+  struct list_t *next;
+} list_t;
+
+// true iff list of len nodes with values in [-10,10]
+bool is_list(list_t *l, size_t len)
+{
+  if(len == 0)
+    return l == NULL;
+  else
+    return __CPROVER_is_fresh(l, sizeof(*l)) && -10 <= l->value &&
+           l->value <= 10 && is_list(l->next, len - 1);
+}
+```
+
+One can also simply describe finite nested structures:
+
+```c
+typedef struct buffer_t
+{
+  size_t size;
+  char *arr;
+  char *cursor;
+} buffer_t;
+
+typedef struct double_buffer_t
+{
+  buffer_t *first;
+  buffer_t *second;
+} double_buffer_t;
+
+bool is_sized_array(char *arr, size_t size)
+{
+  return __CPROVER_is_fresh(arr, size);
+}
+
+bool is_ranged_cursor(char *cursor, char *arr, size_t size)
+{
+  void *lb = arr;
+  void *ub = arr + size;
+  __CPROVER_is_in_range(lb, cursor, ub);
+}
+
+bool is_buffer(buffer_t *b)
+{
+  return __CPROVER_is_fresh(b, sizeof(*b)) && (0 < b->size && b->size <= 10) &&
+         is_sized_array(b->arr, b->size) &&
+         is_ranged_cursor(b->cursor, b->arr, b->size);
+}
+
+bool is_double_buffer(double_buffer_t *b)
+{
+  return __CPROVER_is_fresh(b, sizeof(*b)) && is_buffer(b->first) &&
+         is_buffer(b->second);
+}
+```
+
+And one can then use these predicates in requires or ensures clauses for function
+contracts.
+
+```c
+int foo(list_t *l, double_buffer_t *b)
+  // clang-format off
+  __CPROVER_requires(is_list(l, 3))
+  __CPROVER_requires(is_double_buffer(b))
+  __CPROVER_ensures(-28 <= __CPROVER_return_value &&
+                    __CPROVER_return_value <= 50)
+// clang-format on
+{
+  // access the assumed data structure
+  return l->value + l->next->value + l->next->next->value + b->first->size +
+         b->second->size;
+}
+```
+
+### Limitations
+
+The main limitation with user defined predicates is that they must be bounded.
+In the `is_list` example above, recursion is bounded by the use of the `len`
+parameter.
+The `is_double_buffer` predicate also describes a bounded structure.
```

Comment:
Do we report any errors in case the predicate is not bounded or CBMC will just try to unroll it forever.

Reply:
For now it is in the user's hands to ensure boundedness. Symex will just run forever and the user will have to ctrl-c it.

Reply:
We could later impose syntactic restrictions like guarded recursion and implement a lazy expansion of the predicates (automatically upon access or using open/close statements like in verifast or similar tools).